### DISCUSSION TITLE
Handling the Quit event on Mac

### DISCUSSION THREAD

**Discussion by miguelgrinberg** [fltk/fltk#1085] [Q&A] [answered]
Hi! I have an application that installs itself in the system tray. On Mac I would like to handle two different types of close event:

- If the user closes the FLTK window I want keep the application running just with the tray icon
- If the user selects Quit from the Mac application menu added by FLTK (or presses Cmd-Q), I want to close the window as well, but also close the tray icon and exit.

The problem that I have is that I can't find a way to tell if a close event on the main window was caused by a close event on the window or by the Mac's quit event. Any suggestions on how to tell these two events apart? Thanks!

**Comment by Albrecht-S** [comment]
Unfortunately I can't help with these macOS specific questions, but maybe our macOS expert @ManoloFLTK can help (mentioning him here so he gets notified).

**Reply by ManoloFLTK** [reply, marked as answer]
I don't know what is the tray icon, but I believe FLTK 1.4 has the solution for you with function`Fl::program_should_quit()`. Consider this code as an example to continue the program after the window is closed and to stop it only when the Quit menu command has been issued:
```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Box.H>

int main(int argc, char **argv) {
  Fl_Window *window = new Fl_Window(340, 180);
  Fl_Box *box = new Fl_Box(20, 40, 300, 100, "Hello, World!");
  box->box(FL_UP_BOX);
  box->labelsize(36);
  window->end();
  window->show();
  while (!Fl::program_should_quit()) Fl::wait(1000);
  return 0;
}
```

**Reply by miguelgrinberg** [reply]
@ManoloFLTK Thanks for your answer. The tray icon is implemented separately from FLTK, obviously, so it is mostly unrelated to the question. I also have no problem keeping the application running without any windows open and am already using `Fl::wait()` as you indicate in your example.

The question is about knowing when to exit. The `Fl::program_should_quit()` function returns true when the last window is closed. I do not want to exit at that time. I want to exit only when the user chooses Quit from the Mac application menu. The problem I have is that there seems to be no quit event that is sent to the application. If there are windows open, selecting Quit sends a close event to them, and this event cannot be distinguish from a real close, as opposite to a quit event. Should I assume this cannot be done?

**Reply by ManoloFLTK** [reply]
You write
> The Fl::program_should_quit() function returns true when the last window is closed. 

which suggests you did not try the example program I gave you. May be you should.

**Reply by miguelgrinberg** [reply]
My apologies. My reading of the source code of the `program_should_quit()` function was incorrect. The function does return true only after the Quit command is triggered. Thanks!

The remaining issue is that `program_should_quit()` returns `false` while the window(s) are being forcefully closed as a result of the user quitting the app, so my application cannot react to the quit event before my window gets closed. For example, I would like to show a confirmation prompt before exiting, allowing the user to abort and return to the app, but by the time my application sees `program_should_quit() == true` the window has already been closed.

Would you be open to move the `program_should_quit(1)` above the loop that closes the windows, so that the application can check for it in the window close event? Something like this:

```objc
- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication*)sender
{
  Fl::program_should_quit(1);  // apps can check this flag while handling close events
  fl_lock_function();
  while ( Fl_X::first ) {
    Fl_Window *win = Fl::first_window();
    if (win->parent()) win = win->top_window();
    Fl_Widget_Tracker wt(win); // track the window object
    Fl::handle(FL_CLOSE, win);
    if (wt.exists() && win->shown()) { // the user didn't close win
      break;
    }
  }
  fl_unlock_function();
  if ( ! Fl::first_window() ) {
    Fl_Cocoa_Screen_Driver::breakMacEventLoop(); // necessary when called through menu and in Fl::wait()
  } else {
    Fl::program_should_quit(0);  // one or more windows are still shown, abort quit
  }
  return NSTerminateCancel;
}
```

Happy to submit a PR with this change if it makes sense!

**Reply by ManoloFLTK** [reply]
Why don't you put the confirmation dialog in the window callback function that runs when the user attempts to close the window?

**Reply by miguelgrinberg** [reply]
I am planning to do that, that is actually fine. But I only want to show the confirmation alert when the user is quitting, not when the user is closing the window. Not sure if I was clear before, the value of `program_should_quit()` is 0 in the window callback because the window is closed before this flag is set to 1.

**Reply by ManoloFLTK** [reply]
I've considered in detail your proposal to slightly change the FLTK implementation of method `applicationShouldTerminate:`. That would create a subtle problem that `Fl::application_should_quit()` would become
`Returns non-zero when a request for program termination was received.`
instead of what it is now
`Returns non-zero when a request for program termination was received and accepted. `

My proposal to you is therefore to use a bit a Objective-C in your program as in the following example program in order to reimplement `applicationShouldTerminate:` according to your needs:
```
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Box.H>
#include <FL/fl_ask.H>

#import <Cocoa/Cocoa.h>

bool miguel_program_should_quit = false;

@interface FLAppDelegate : NSObject <NSApplicationDelegate> // Implemented in libfltk
@end

@interface miguelAppDelegate : FLAppDelegate
- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication*)sender;
@end

@implementation miguelAppDelegate
- (NSApplicationTerminateReply)applicationShouldTerminate:(NSApplication*)sender {
  miguel_program_should_quit = true;
  NSApplicationTerminateReply retval = [super applicationShouldTerminate:sender];
  if (Fl::first_window()) miguel_program_should_quit = false;
  return retval;
}
@end

void change_delegate() {
  [NSApp setDelegate:[[miguelAppDelegate alloc] init]];
}

void close_cb(Fl_Widget *w) {
  if (!miguel_program_should_quit || fl_choice("close?", "No", "Yes", NULL)) delete w;
}


int main(int argc, char **argv) {
  Fl_Window *window = new Fl_Window(340, 180);
  Fl_Box *box = new Fl_Box(20, 40, 300, 100, "Hello, World!");
  box->box(FL_UP_BOX);
  box->labelsize(36);
  window->end();
  window->callback(close_cb);
  window->show();
  change_delegate();
  while (!miguel_program_should_quit) Fl::wait(1000);
  return 0;
}
```

**Reply by miguelgrinberg** [reply]
I appreciate the suggestion, but this would be too complicated for me, as I'm using FLTK via the Rust bindings. Much easier to compile fltk-rs with my fork of fltk that implements the change I suggested above.

I think my use case is not unusual, though. I do understand that the documentation of `program_should_quit()` would have to be updated and maybe that is not something you want to do. In that case consider adding a different function to report a quit request.

In any case, thanks for all the help.
